core/src/main/java/org/phoenixctms/ctsms/service/proband/ProbandServiceImpl.java

```diff
@@ -2905,4 +2912,150 @@ private void notifyProbandInactive(ProbandStatusEntry statusEntry, Date now) thr
 			}
 		}
 	}
+
+	@Override
+	protected ProbandOutVO handleUpdateProbandDepartment(AuthenticationVO auth, Long probandId, Long newDepartmentId, String plainNewDepartmentPassword,
+			String plainOldDepartmentPassword) throws Exception {
+		ProbandDao probandDao = this.getProbandDao();
+		Proband proband = CheckIDUtil.checkProbandId(probandId, probandDao, LockMode.PESSIMISTIC_WRITE);
+		Department newDepartment = CheckIDUtil.checkDepartmentId(newDepartmentId, this.getDepartmentDao());
+		if (plainNewDepartmentPassword == null) {
+			plainNewDepartmentPassword = CoreUtil.getUserContext().getPlainDepartmentPassword();
+		}
+		if (plainOldDepartmentPassword == null) {
+			plainOldDepartmentPassword = CoreUtil.getUserContext().getPlainDepartmentPassword();
+		}
+		Department oldDepartment = proband.getDepartment();
+		if (!oldDepartment.equals(newDepartment)) {
+			if (!CryptoUtil.checkDepartmentPassword(newDepartment, plainNewDepartmentPassword)) {
+				throw L10nUtil.initServiceException(ServiceExceptionCodes.DEPARTMENT_PASSWORD_WRONG);
+			}
+			if (!CryptoUtil.checkDepartmentPassword(oldDepartment, plainOldDepartmentPassword)) {
+				throw L10nUtil.initServiceException(ServiceExceptionCodes.OLD_DEPARTMENT_PASSWORD_WRONG);
+			}
+			Timestamp now = new Timestamp(System.currentTimeMillis());
```

Comment:
What happens if an exception occurs in one of the re-encryptors?

Reply:
the transaction is abandoned instantly. this also holds for the re-encryption of large files.